src/core/image.rs

```diff
@@ -81,6 +86,34 @@ where
     fn expose_ports(&self) -> Vec<u16> {
         Default::default()
     }
+
+    /// Returns the commands that needs to be executed after a container is started i.e. commands
+    /// to be run in a running container.
+    ///
+    /// This method is useful when certain re-configuration is required after the start
+    /// of container for the container to be considered ready for use in tests.
+    fn exec_after_start(&self) -> Option<Vec<ExecCommand>> {
```

Comment:
What do you think of this function taking an instance of the `ContainerDetails` type?

Then an implementation could search for whatever data it needs and we would not have to maintain the `Substitution` abstraction and extend it to allow for other usecases than figuring out the host port for example.

I guess it is a trade-off between how many details we want to expose.

Overall, I would prefer if we pass data _into_ this function and let the function figure out how to construct the desired command than passing a template out.

We can either:

- use `ContainerDetails` and accept that this type is the part of our public API
- create our own type and keep things limited for now (to for example ports)

I don't mind much either way with a slight preference for the latter.

Adding our own type allows us to extend it with convenience functions like extracing a port mapping.

Reply:
Do we need the `Option<Vec>`? What is the semantic difference between `None` and `Some(empty-vec)`?